File: 1-two-sum/1-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_dict = {}
        for i in range(len(nums)):
            if num_dict.get(nums[i], None) is None:
                num_dict[nums[i]] = []

            num_dict[nums[i]].append(i)
            
        for i in range(len(nums)):
            diff_pos = num_dict.get(target - nums[i], None)
            if diff_pos is not None:
                for pos in diff_pos:
                    if pos != i:
                        return [i, pos]
                    
        return None

File: 1-two-sum/1-two-sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    num_hash = nums.each.with_index.each_with_object({}) do |(num, idx), hash|
        hash[num] ||= []
        hash[num].push(idx)
    end
    
    result = nil
    nums.each.with_index do |num, idx|
        diff_pos = num_hash[target - num] || []
        other_pos = diff_pos.detect { |pos| pos != idx }
        result = [idx, other_pos] unless other_pos.nil?
        break unless result.nil?
    end
    
    result
end